### PR TITLE
Remove `skipLibCheck` from most workspaces

### PR DESCRIPTION
**Pull request by kachkaev**
Enabled by #577, [Asana task](https://app.asana.com/0/0/1202922733837195) (internal)

This PR removes [`skipLibCheck`](https://www.typescriptlang.org/tsconfig#skipLibCheck) from `@local/tsconfig`. This makes `yarn lint:tsc` more rigorous, which helps us spot issues in `.d.ts` files (mostly in `node_modules`). That can guard us from runtime bugs, because with `skipLibCheck` we may get `any` from some library typings withour realising that.

TSC without `skipLibCheck` are a bit slower, so I’ve introduced `--skipLibCheck` argument in build scripts, where applicable. This makes cold `yarn install` a bit faster (≈25s → ≈15s). Type integrity is still validated via `lint:tsc`.

I had to fix several TypeScript issues for CI to pass – you can find details of these changes in PR comments. The only workspace I have left with `skipLibCheck` for now is `site`. It still have quite a few issues:

<details>
<summary><code>Found 38 errors in 24 files.</code></summary>
<pre>
../node_modules/mdast-util-mdx-expression/complex-types.d.ts:35:5 - error TS2300: Duplicate identifier 'mdxFlowExpression'.

35     mdxFlowExpression: MdxFlowExpression
       ~~~~~~~~~~~~~~~~~

../node_modules/mdast-util-mdx-expression/complex-types.d.ts:36:5 - error TS2300: Duplicate identifier 'mdxFlowExpression'.

36     mdxFlowExpression: MdxFlowExpression
       ~~~~~~~~~~~~~~~~~

../node_modules/mongodb/mongodb.d.ts:1744:97 - error TS2344: Type 'TLocal' does not satisfy the constraint 'Document'.

1744     watch<TLocal = TSchema>(pipeline?: Document[], options?: ChangeStreamOptions): ChangeStream<TLocal>;
                                                                                                     ~~~~~~

  ../node_modules/mongodb/mongodb.d.ts:1744:11
    1744     watch<TLocal = TSchema>(pipeline?: Document[], options?: ChangeStreamOptions): ChangeStream<TLocal>;
                   ~~~~~~~~~~~~~~~~
    This type parameter might need an `extends Document` constraint.

../node_modules/mongodb/mongodb.d.ts:2536:99 - error TS2344: Type 'TSchema' does not satisfy the constraint 'Document'.

2536     watch<TSchema = Document>(pipeline?: Document[], options?: ChangeStreamOptions): ChangeStream<TSchema>;
                                                                                                       ~~~~~~~

  ../node_modules/mongodb/mongodb.d.ts:2536:11
    2536     watch<TSchema = Document>(pipeline?: Document[], options?: ChangeStreamOptions): ChangeStream<TSchema>;
                   ~~~~~~~~~~~~~~~~~~
    This type parameter might need an `extends Document` constraint.

../node_modules/mongodb/mongodb.d.ts:3309:5 - error TS2416: Property 'end' in type 'GridFSBucketWriteStream' is not assignable to the same property in base type 'Writable'.
  Type '{ (): void; (chunk: Buffer): void; (callback: Callback<void | GridFSFile>): void; (chunk: Buffer, callback: Callback<void | GridFSFile>): void; (chunk: Buffer, encoding: BufferEncoding): void; (chunk: Buffer, encoding: BufferEncoding | undefined, callback: Callback<...>): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

3309     end(): void;
         ~~~

../node_modules/mongodb/mongodb.d.ts:3310:5 - error TS2416: Property 'end' in type 'GridFSBucketWriteStream' is not assignable to the same property in base type 'Writable'.
  Type '{ (): void; (chunk: Buffer): void; (callback: Callback<void | GridFSFile>): void; (chunk: Buffer, callback: Callback<void | GridFSFile>): void; (chunk: Buffer, encoding: BufferEncoding): void; (chunk: Buffer, encoding: BufferEncoding | undefined, callback: Callback<...>): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.

3310     end(chunk: Buffer): void;
         ~~~

../node_modules/mongodb/mongodb.d.ts:3311:5 - error TS2416: Property 'end' in type 'GridFSBucketWriteStream' is not assignable to the same property in base type 'Writable'.
  Type '{ (): void; (chunk: Buffer): void; (callback: Callback<void | GridFSFile>): void; (chunk: Buffer, callback: Callback<void | GridFSFile>): void; (chunk: Buffer, encoding: BufferEncoding): void; (chunk: Buffer, encoding: BufferEncoding | undefined, callback: Callback<...>): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.

3311     end(callback: Callback<GridFSFile | void>): void;
         ~~~

../node_modules/mongodb/mongodb.d.ts:3312:5 - error TS2416: Property 'end' in type 'GridFSBucketWriteStream' is not assignable to the same property in base type 'Writable'.
  Type '{ (): void; (chunk: Buffer): void; (callback: Callback<void | GridFSFile>): void; (chunk: Buffer, callback: Callback<void | GridFSFile>): void; (chunk: Buffer, encoding: BufferEncoding): void; (chunk: Buffer, encoding: BufferEncoding | undefined, callback: Callback<...>): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.

3312     end(chunk: Buffer, callback: Callback<GridFSFile | void>): void;
         ~~~

../node_modules/mongodb/mongodb.d.ts:3313:5 - error TS2416: Property 'end' in type 'GridFSBucketWriteStream' is not assignable to the same property in base type 'Writable'.
  Type '{ (): void; (chunk: Buffer): void; (callback: Callback<void | GridFSFile>): void; (chunk: Buffer, callback: Callback<void | GridFSFile>): void; (chunk: Buffer, encoding: BufferEncoding): void; (chunk: Buffer, encoding: BufferEncoding | undefined, callback: Callback<...>): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.

3313     end(chunk: Buffer, encoding: BufferEncoding): void;
         ~~~

../node_modules/mongodb/mongodb.d.ts:3314:5 - error TS2416: Property 'end' in type 'GridFSBucketWriteStream' is not assignable to the same property in base type 'Writable'.
  Type '{ (): void; (chunk: Buffer): void; (callback: Callback<void | GridFSFile>): void; (chunk: Buffer, callback: Callback<void | GridFSFile>): void; (chunk: Buffer, encoding: BufferEncoding): void; (chunk: Buffer, encoding: BufferEncoding | undefined, callback: Callback<...>): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.

3314     end(chunk: Buffer, encoding: BufferEncoding | undefined, callback: Callback<GridFSFile | void>): void;
         ~~~

../node_modules/mongodb/mongodb.d.ts:3996:99 - error TS2344: Type 'TSchema' does not satisfy the constraint 'Document'.

3996     watch<TSchema = Document>(pipeline?: Document[], options?: ChangeStreamOptions): ChangeStream<TSchema>;
                                                                                                       ~~~~~~~

  ../node_modules/mongodb/mongodb.d.ts:3996:11
    3996     watch<TSchema = Document>(pipeline?: Document[], options?: ChangeStreamOptions): ChangeStream<TSchema>;
                   ~~~~~~~~~~~~~~~~~~
    This type parameter might need an `extends Document` constraint.

../node_modules/next-seo/lib/jsonld/article.d.ts:2:31 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

2 import { ArticleAuthor } from 'src/types';
                                ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/brand.d.ts:1:38 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

1 import type { AggregateRating } from 'src/types';
                                       ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/breadcrumb.d.ts:1:39 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

1 import type { ItemListElements } from 'src/types';
                                        ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/carousel.d.ts:2:59 - error TS2307: Cannot find module 'src/index' or its corresponding type declarations.

2 import type { CourseJsonLdProps, RecipeJsonLdProps } from 'src/index';
                                                            ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/carousel.d.ts:3:46 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

3 import type { Review, AggregateRating } from 'src/types';
                                               ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/collectionPage.d.ts:2:35 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

2 import type { CreativeWork } from 'src/types';
                                    ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/corporateContact.d.ts:2:35 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

2 import type { ContactPoint } from 'src/types';
                                    ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/course.d.ts:1:31 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

1 import type { Provider } from 'src/types';
                                ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/event.d.ts:2:111 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

2 import type { Location, AggregateOffer, Offers, Performer, Organizer, EventStatus, EventAttendanceMode } from 'src/types';
                                                                                                                ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/faqPage.d.ts:1:31 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

1 import type { Question } from 'src/types';
                                ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/howTo.d.ts:2:42 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

2 import { PriceSpecification, Step } from 'src/types';
                                           ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/localBusiness.d.ts:2:113 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

2 import type { Address, Geo, AggregateRating, Review, Action, GeoCircle, OpeningHoursSpecification, Offer } from 'src/types';
                                                                                                                  ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/organization.d.ts:1:66 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

1 import type { Address, OrganizationCategory, ContactPoint } from 'src/types';
                                                                   ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/product.d.ts:2:70 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

2 import type { Review, AggregateRating, AggregateOffer, Offers } from 'src/types';
                                                                       ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/profilePage.d.ts:2:39 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

2 import type { ItemListElements } from 'src/types';
                                        ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/qaPage.d.ts:1:31 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

1 import type { Question } from 'src/types';
                                ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/recipe.d.ts:2:58 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

2 import type { Instruction, AggregateRating, Video } from 'src/types';
                                                           ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/softwareApp.d.ts:2:46 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

2 import type { Review, AggregateRating } from 'src/types';
                                               ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/video.d.ts:1:23 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

1 import { Video } from 'src/types';
                        ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/videoGame.d.ts:1:96 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

1 import type { Offers, AggregateRating, GamePlayMode, Video, Review, ApplicationCategory } from 'src/types';
                                                                                                 ~~~~~~~~~~~

../node_modules/next-seo/lib/jsonld/webPage.d.ts:1:33 - error TS2307: Cannot find module 'src/types' or its corresponding type declarations.

1 import type { ReviewedBy } from 'src/types';
                                  ~~~~~~~~~~~

../node_modules/next/dist/client/with-router.d.ts:8:121 - error TS2344: Type 'C' does not satisfy the constraint 'BaseContext'.

8 export default function withRouter<P extends WithRouterProps, C = NextPageContext>(ComposedComponent: NextComponentType<C, any, P>): React.ComponentType<ExcludeRouterProps<P>>;
                                                                                                                          ~

  ../node_modules/next/dist/client/with-router.d.ts:8:63
    8 export default function withRouter<P extends WithRouterProps, C = NextPageContext>(ComposedComponent: NextComponentType<C, any, P>): React.ComponentType<ExcludeRouterProps<P>>;
                                                                    ~~~~~~~~~~~~~~~~~~~
    This type parameter might need an `extends BaseContext` constraint.

theme-override.d.ts:71:5 - error TS2717: Subsequent property declarations must have the same type.  Property 'grey' must be of type 'Color', but here has type 'undefined'.

71     grey: undefined;
       ~~~~

  ../node_modules/@mui/material/styles/createPalette.d.ts:92:3
    92   grey: Color;
         ~~~~
    'grey' was also declared here.

theme-override.d.ts:139:5 - error TS2300: Duplicate identifier 'warning'.

139     warning: true;
        ~~~~~~~

theme-override.d.ts:148:5 - error TS2300: Duplicate identifier 'warning'.

148     warning: false;
        ~~~~~~~

theme-override.d.ts:148:5 - error TS2717: Subsequent property declarations must have the same type.  Property 'warning' must be of type 'true', but here has type 'false'.

148     warning: false;
        ~~~~~~~

  theme-override.d.ts:139:5
    139     warning: true;
            ~~~~~~~
    'warning' was also declared here.

theme-override.d.ts:162:16 - error TS2714: The expression of an export assignment must be an identifier or qualified name in an ambient context.

162 export default "";
                   ~~

Found 38 errors in 24 files.

Errors  Files
     2  ../node_modules/mdast-util-mdx-expression/complex-types.d.ts:35
     9  ../node_modules/mongodb/mongodb.d.ts:1744
     1  ../node_modules/next-seo/lib/jsonld/article.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/brand.d.ts:1
     1  ../node_modules/next-seo/lib/jsonld/breadcrumb.d.ts:1
     2  ../node_modules/next-seo/lib/jsonld/carousel.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/collectionPage.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/corporateContact.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/course.d.ts:1
     1  ../node_modules/next-seo/lib/jsonld/event.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/faqPage.d.ts:1
     1  ../node_modules/next-seo/lib/jsonld/howTo.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/localBusiness.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/organization.d.ts:1
     1  ../node_modules/next-seo/lib/jsonld/product.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/profilePage.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/qaPage.d.ts:1
     1  ../node_modules/next-seo/lib/jsonld/recipe.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/softwareApp.d.ts:2
     1  ../node_modules/next-seo/lib/jsonld/video.d.ts:1
     1  ../node_modules/next-seo/lib/jsonld/videoGame.d.ts:1
     1  ../node_modules/next-seo/lib/jsonld/webPage.d.ts:1
     1  ../node_modules/next/dist/client/with-router.d.ts:8
     5  theme-override.d.ts:71
error Command failed with exit code 1.
</pre>
</details>

We can look into them separately one day.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,13 +44,13 @@
     ]
   },
   "resolutions": {
-    "@types/react": "^18.0.0"
+    "@types/react": "^18.0.15"
```
Comment (kachkaev) [inline package.json:47]: Result of `yarn fix:dependency-version-consistency` (v3)

File: package.json [hunk 2]
```diff
@@ -24,7 +29,7 @@
     "lint:lockfile-lint": "lockfile-lint --path yarn.lock --allowed-hosts registry.yarnpkg.com --allowed-schemes \"https:\"",
     "lint:markdownlint": "markdownlint \"**/*\"",
     "lint:prettier": "prettier --cache --check --ignore-unknown \"**/*\"",
-    "lint:tsc": "turbo run lint:tsc",
+    "lint:tsc": "turbo run --continue lint:tsc",
```
Comment (kachkaev) [inline package.json:32]: Helps gather more feedback from single run

File: site/package.json [hunk 3]
```diff
@@ -31,8 +31,8 @@
     "@fortawesome/free-regular-svg-icons": "^6.0.0",
     "@fortawesome/free-solid-svg-icons": "^6.0.0",
     "@istanbuljs/nyc-config-typescript": "1.0.2",
-    "@mui/icons-material": "^5.8.4",
-    "@mui/material": "^5.10.1",
+    "@mui/icons-material": "^5.10.3",
+    "@mui/material": "^5.10.4",
```
Comment (kachkaev) [inline site/package.json:35]: Fixes a few issues with typings

File: site/package.json [hunk 4]
```diff
@@ -120,6 +120,7 @@
     "@types/react-html-parser": "^2.0.2",
     "@types/react-slick": "^0.23.7",
     "@types/tar": "^6.1.2",
+    "@types/tmp": "^0.2.3",
```
Comment (kachkaev) [inline site/package.json:123]: Needed to fix missing typings in `tmp-promise` (if I remember package name correctly)

File: packages/@blockprotocol/type-system-node/tsconfig.json [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {
-  "extends": "@local/tsconfig/node.json",
+  "extends": "@local/tsconfig/dom.json",
```
Comment (kachkaev) [inline packages/@blockprotocol/type-system-node/tsconfig.json:2]: Fixes missing `WebAssembly` in packages/@blockprotocol/type-system-{node,web}dist/index.d.ts`

File: packages/@blockprotocol/type-system-web/tsconfig.json [hunk 6]
```diff
@@ -1,5 +1,5 @@
 {
-  "extends": "@local/tsconfig/node.json",
+  "extends": "@local/tsconfig/dom.json",
```
Comment (kachkaev) [inline packages/@blockprotocol/type-system-web/tsconfig.json:2]: Fixes missing `WebAssembly` in packages/@blockprotocol/type-system-{node,web}dist/index.d.ts`

File: packages/@local/tsconfig/dom.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (kachkaev) [inline packages/@local/tsconfig/dom.json:1]: Adds flexibility: we can inherit non-React tsconfig with DOM

File: package.json [hunk 8]
```diff
@@ -2,13 +2,18 @@
...
+  "workspaces": {
+    "packages": [
+      "crates/*",
+      "packages/*",
+      "packages/@blockprotocol/*",
+      "packages/@local/*",
+      "site"
+    ],
+    "nohoist": [
+      "**/@types/prismjs"
```
Comment (kachkaev) [inline package.json:14]: Prismjs typings refer to DOM nodes, so when they are hoisted, workspaces with node-only tsconfig fail with this error:

```
../../node_modules/@types/prismjs/index.d.ts(73,46): error TS2304: Cannot find name 'ParentNode'.
```

File: packages/@local/tsconfig/node.json [hunk 9]
```diff
@@ -14,7 +14,6 @@
     "moduleResolution": "Node",
     "noUncheckedIndexedAccess": true,
     "resolveJsonModule": true,
-    "skipLibCheck": true,
```
Comment (kachkaev) [inline packages/@local/tsconfig/node.json:17]: 🎉

File: packages/blockprotocol/cli/commands/publish/post-form.js [hunk 10]
```diff
@@ -20,13 +20,15 @@ export const postPublishForm = async ({
...
 
+  /** @type {import("node-fetch").FormData} -- https://github.com/node-fetch/node-fetch/issues/900#issuecomment-716342574 */
   const form = new FormData();
 
   const file = await fileFromPath(tarballFilePath);
 
   form.set("blockName", blockName);
   form.set("tarball", file, "tarball.tar.gz");
 
+  /** @type {import("node-fetch").RequestInit} */
```
Comment (kachkaev) [inline packages/blockprotocol/cli/commands/publish/post-form.js:31]: Helps spot problems earlier (before the object is passed to `fetch` as an arg)

File: packages/@local/tsconfig/dom.json (hunk 7 above)
Comment (CiaranMn) [reply packages/@local/tsconfig/dom.json:1]: Since we have this, we could use it in `@blockprotocol/*`, since none of them contain JSX

**Review by CiaranMn — APPROVED** [review]

File: packages/@local/tsconfig/dom.json (hunk 7 above)
Comment (kachkaev) [reply packages/@local/tsconfig/dom.json:1]: Good idea! https://github.com/blockprotocol/blockprotocol/pull/585/commits/36f6bba8d3ae285699e668a5cada77990cf5ed79

File: package.json [hunk 11]
```diff
@@ -38,14 +43,11 @@
       "suppress-exit-code prettier --write"
     ]
   },
-  "resolutions": {
-    "@types/react": "^18.0.0"
-  },
```
Comment (kachkaev) [inline package.json:43]: Fixes `yarn lint:dependency-version-consistency"

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**